go/runtime/scheduling/simple/txpool/priorityqueue/priority_queue.go

```diff
@@ -187,6 +192,57 @@ func (q *priorityQueue) removeTxsLocked(items []*item) {
 	}
 }
 
+// Implements api.TxPool.
+func (q *priorityQueue) GetPrioritizedBatch(offset *hash.Hash, limit uint32) []*transaction.CheckedTransaction {
+	q.Lock()
+	defer q.Unlock()
+
+	var (
+		batch      []*transaction.CheckedTransaction
+		toRemove   []*item
+		offsetItem btree.Item
+	)
+	if offset != nil {
+		var exists bool
+		offsetItem, exists = q.transactions[*offset]
+		if !exists {
+			// Offset does not exist so no items will be matched anyway.
+			return nil
+		}
+		limit++ // Since the offset item will be included.
+	}
+	q.priorityIndex.DescendLessOrEqual(offsetItem, func(i btree.Item) bool {
+		item := i.(*item)
+
+		for w, l := range q.weightLimits {
+			txW := item.tx.Weight(w)
+			// Transaction weight greater than the limit. Drop the tx from the pool.
+			if txW > l {
+				toRemove = append(toRemove, item)
+				return true
+			}
+		}
+
+		// Add the tx to the batch.
+		batch = append(batch, item.tx)
+		if uint32(len(batch)) >= limit { //nolint: gosimple
+			return false
+		}
+		return true
+	})
+	if offset != nil {
+		// Skip the offset item itself.
+		batch = batch[1:]
```

Comment:
latecoming comment, sorry:

it's harder to tell that this is correct, compared to having an equality check inside the loop

- could the weight limits have changed, to make us skip the offset hash?
- could there have been duplicates of the offset hash?
- could something else have removed the offset hash?

some of these questions may have safe answers for how we use this data structure in the scheduler. it would be nice if we could make it safe in the general case

Reply:
Ah you're right, let's fix that, can you comment in the non-backport PR?

In any case this could only happen if: i) the offset tx was included at initial query time, ii) weight limits have changed (this implies an epoch transition as they cannot otherwise change), and iii) offset was used in the subsequent query. Since an epoch transition will cause processing to be aborted anyway, this is not that bad. But it should still be fixed (by adding an equality check or just not removing these txes here).

Reply:
:+1: https://github.com/oasisprotocol/oasis-core/pull/4438/files#r793924684